[jest-expo]: Use Jest 26 instead of 25.

chrisbobbe opened the pull request:
Based on the example in 6c26f5319 (#7556). Please feel free to close this if this is being tracked somewhere else, but I didn't find it anywhere.

### Test plan
- [x] `yarn test` in packages/jest-expo/package.json passes
- [ ] I'm not sure if anything else needs to be tested, or if other things would need to update to use Jest 26 along with this.

-----

### Motivation

I have an existing React Native app that uses a few Expo modules via `react-native-unimodules`. I find it convenient to use `jest-expo` so that I don't have to mock those modules myself. Also, [the `jest-expo/ios` and `jest-expo/android` presets](https://github.com/expo/expo/tree/master/packages/jest-expo#platforms) look very helpful, and I hope to experiment with them. Thanks for those! 🎉

My project needs to use Jest 26 because it needs to use [Jest's modern fake-timer implementation](https://github.com/facebook/jest/blob/master/CHANGELOG.md#2600), which is new in 26.

When I use jest-expo at its latest, Jest 25 is getting pulled in instead of 26, and modern fake timers aren't available. When I add the following to my package.json, Jest 26 gets pulled in and everything seems fine:

```json
  "resolutions": {
    "jest-expo/jest": "^26.4.1"
  }
```

But I think that's not the most correct solution; eventually, `jest-expo` itself should be updated, as this PR aims to do, to pull in Jest 26. I'd expect Yarn to give me a warning with the above `resolutions` code, because `^26.4.1` is entirely outside the original range, `^25.2.0`. From the [doc](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/#toc-tips-tricks): "You will receive a warning if your resolution version or range is not compatible with the original version range." I haven't seen a warning like that yet; that sounds like a possible Yarn bug; that doc says `resolutions` is a fairly new feature where some things might not work.

chrisbobbe committed:
[jest-expo]: Use Jest 26 instead of 25.

Done by emulating the changes to packages/jest-expo/package.json in
6c26f5319 (#7556):

- Upgrade `jest` and `babel-jest` to the latest 26.x (as of writing).

- No need to upgrade `jest-watch-select-projects`, though; 2.0.0 is
  still the current latest:
    https://github.com/jest-community/jest-watch-select-projects/blob/eb4f3c070/package.json#L3

- Upgrade `jest-watch-typeahead` to 0.6.2 [1]. The only breaking change
  announced is that Node 8 support is dropped [2]; specifically,
  Node 10 is now the minimum supported. This should be fine: Expo's
  policy is to work with "Node.js LTS release or greater" [3], and
  pre-v10 releases are not LTS [4]. (In fact, even v10 itself isn't
  LTS anymore.)

[1] We're hesitating to allow 0.6.4; for why, please see
    https://github.com/expo/expo/pull/12735#issuecomment-876789269.
[2] https://github.com/jest-community/jest-watch-typeahead/blob/master/CHANGELOG.md#060
[3] https://docs.expo.io/get-started/installation/#requirements
[4] https://nodejs.org/en/about/releases/

chrisbobbe commented:
(Just resolved a merge conflict.)

chrisbobbe commented:
(@SimenB, hopefully this is helpful, but I'd understand if you want to lead this work yourself. 🙂)

SimenB commented:
I'm a big fan 👍👍👍 Any reason not to go to 27

brentvatne reviewed (changes requested):
1. it looks like you didn't run `yarn`, so the lockfile didn't update
2. jest-watch-typeahead depends on a more recent version of jest-watcher, we should make sure it uses a version that is compatible with jest ^26:

```
=> Found "jest-watch-typeahead#jest-watcher@27.0.6"
info This module exists because "_project_#jest-expo#jest-watch-typeahead" depends on it.
```

chrisbobbe committed:
[jest-expo] Run yarn from packages/jest-expo.

I think I'd have expected this not to change the lockfile. Maybe a
previous commit forgot to run `yarn`?

chrisbobbe commented:
Thanks for the review!

> 2\. jest-watch-typeahead depends on a more recent version of jest-watcher, we should make sure it uses a version that is compatible with jest ^26:
> 
> ```
> => Found "jest-watch-typeahead#jest-watcher@27.0.6"
> info This module exists because "_project_#jest-expo#jest-watch-typeahead" depends on it.
> ```

Thanks for catching this.

Squinting a bit, it looks like jest-watch-typeahead@^0.6.4 technically declares itself compatible with Jest 26. In jest-community/jest-watch-typeahead@6552ead2d, released in jest-watch-typeahead@0.6.4, the jest peer-dep range is changed from ^26.0.0 to ^26.0.0 || ^27.0.0 (not dropping 26), and I didn't see any peer-dep warnings from jest-watch-typeahead@0.6.4's own dependencies (including jest-watcher@27.0.6, which [doesn't have any peer deps at all](https://github.com/facebook/jest/blob/v27.0.6/packages/jest-watcher/package.json)).

Still, I agree that the "27" looks a little suspicious in a switch to Jest 26, so I've bumped jest-watch-typeahead down from ^0.6.4 to 0.6.2 (no caret) and noted it in the commit message. 0.6.2 is before 0.6.4, so it shouldn't have jest-community/jest-watch-typeahead@6552ead2d, and 0.6.3 (which also shouldn't have it) didn't have a changelog entry, so, just to be safe.

I ran `yarn` in the commit that changes packages/jest-expo/package.json, and kept the lockfile changes. I ran `yarn test` again.

(I also ran `yarn` from packages/jest-expo/package.json in a commit just before that; it looks like a previous commit forgot to commit some lockfile changes?)

extrct commented:
+1 for bumping to ^27.0.6 and a cutting a major release. Solves a setup snag adding jest to fresh expo projects, detailed in [this PR](https://github.com/expo/expo/pull/11580) and potentially addresses [this issue](https://github.com/expo/expo/issues/9541)

chrisbobbe commented:
> bumping to ^27.0.6 and a cutting a major release

@brentvatne, what do you think? 🙂

brentvatne commented:
this kind of change isn't something we would introduce in the middle of an sdk cycle, but i can release a new minor version with jest 26 if you'd like! how does jest-expo@42.1.0 sound for you?

chrisbobbe commented:
> this kind of change isn't something we would introduce in the middle of an sdk cycle, but i can release a new minor version with jest 26 if you'd like! how does jest-expo@42.1.0 sound for you?

Sounds great! 🙂 Thank you!

I would guess that https://github.com/expo/expo/pull/13549 might want to wait until more coordinated work toward RN v0.64 support in the SDK; is that right? I think I don't really clearly understand when's a good time to make PRs like that one and this one; please let me know if you have tips or a doc I should check. 🙂 Thanks!

brentvatne commented:
@chrisbobbe - we're working on updating react-native to 0.64 for sdk 43 so we could land this once we have bumped the react-native version in this repo. we don't have a good process right now for releasing a version of jest-expo targeting a version of react-native that isn't supported in the sdk, and i'm not sure if want to do that yet

chrisbobbe commented:
Makes sense, thanks for the update!

brentvatne commented:
this is released now in jest-expo@42.1.0

chrisbobbe commented:
Thanks for the update!
